### PR TITLE
Fix sum udaf decimal overflow

### PR DESCRIPTION
**Pull request by bvarghese1**
### Description 

Throw KsqlFunctionException when a Decimal field overflows while using the sum udaf
Addresses https://github.com/confluentinc/ksql/issues/8707

### Testing done 
Added unit and integration(QTT) tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryBuilder.java [hunk 1]
```diff
@@ -641,7 +641,8 @@ private QueryErrorClassifier getConfiguredQueryErrorClassifier(
       final String applicationId
   ) {
     final QueryErrorClassifier userErrorClassifiers = new MissingTopicClassifier(applicationId)
-        .and(new AuthorizationClassifier(applicationId));
+        .and(new AuthorizationClassifier(applicationId))
+        .and(new KsqlFunctionClassifier(applicationId));
```
Comment (jnh5y) [inline ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryBuilder.java:645]: I want to make sure that we get buy-in on the new classifier; I definitely like the idea!

**Review by jnh5y — APPROVED** [review]

**Comment by spena** [comment]
@jnh5y @bvarghese1 I started look at this PR and got wonder why to throw an overflow on the SUM function? A function should be allowed to change the returned schema. 

For instance, 

A query should display the correct sum even if it is higher than `4,1`. The returned value will have a different schema, which may be `10,2` for example, but that should be right. 

There are some rules about the resulted schema in different arithmetic operations:

```
Operation        Result precision                                      Result scale *
e1 + e2             max(s1, s2) + max(p1-s1, p2-s2) + 1        max(s1, s2)
e1 - e2             max(s1, s2) + max(p1-s1, p2-s2) + 1         max(s1, s2)
e1 * e2             p1 + p2 + 1                                                 s1 + s2
e1 / e2             p1 - s1 + s2 + max(6, s1 + p2 + 1)             max(6, s1 + p2 + 1)
e1 % e2           min(p1-s1, p2 -s2) + max( s1,s2 )              max(s1, s2)
```


Here's an example of Mysql:
Other DBs allow the above too. For instance, Mysql:
```
mysql> create table t1(p decimal(2,1));
mysql> insert into t1(p) values(9.99);
mysql> insert into t1(p) values(9.99);
mysql> select sum(p) from t1;
+--------+
| sum(p) |
+--------+
|   19.8  |
+--------+
```

The difference we'll have between other DBs, though, is when using CTAS statements. Other DBs create the resulted table with the schema returned by SUM. For instance:
```
mysql> create table t2 as select sum(p) from t1;
mysql> create table t2 as select sum(p) from t1;
mysql> describe t2;
+--------+---------------+------+-----+---------+-------+
| Field  | Type          | Null | Key | Default | Extra |
+--------+---------------+------+-----+---------+-------+
| sum(p) | decimal(24,1) | YES  |     | NULL    |       |
+--------+---------------+------+-----+---------+-------+
```

Whereas in ksql, the above CTAS does not know what the resulted schema would be until the rows that change the schema are processed. 

I think the option would be to change the Sink table schema dynamically when a new row in in the source table results in a higher decimal schema. Decimal evolution is allowed when precision is higher. Scale can stay the same as that's what the user chooses to do.

@colinhicks Any thoughts?

**Comment by colinhicks** [comment]
Thanks for raising this, @spena, There was some discussion in the corresponding issue: https://github.com/confluentinc/ksql/issues/8707, but it makes sense to give this further thought.

In the CTAS case you're saying there's essentially no way to choose the scale/precision of the output column because the scale of the input is used? And it's not possible to use CAST against the SUM()?

What should happen if I specify the output schema with `value_schema_id`?

**Comment by spena** [comment]
Right, CTAS does not know what the scale and precision would be. Even if you choose a SUM, Ksql uses the schema of the source column.

See:
```
ksql> CREATE TABLE s3 as SELECT ID, sum(value) AS SUM FROM test group by id;
ksql> describe s3;

Name                 : S3
 Field | Type                           
----------------------------------------
 ID    | BIGINT           (primary key) 
 SUM   | DECIMAL(2, 1)
```

Users could use casting if they want to change the sink schema, however this PR still fails on the SUM because it uses the source decimal type to check for overflow instead of the sink.
```
ksql> CREATE TABLE S2 as SELECT ID, CAST(sum(value) AS DECIMAL(4,1)) AS SUM FROM test group by id;
ksql> describe s2;

Name                 : S2
 Field | Type                           
----------------------------------------
 ID    | BIGINT           (primary key) 
 SUM   | DECIMAL(4, 1)                  
----------------------------------------
```

I'm not sure if changing the schema dynamically on the sink is good, though. It's a quick idea it came to my mind, but it needs more thinking. It is the same use case as this user who wanted to pass through all Protobuf fields from source -> sink even if Sink didn't specify them initially. It's part of schema and data evolution.

**Comment by bvarghese1** [comment]
> > Users could use casting if they want to change the sink schema, however this PR still fails on the SUM because it uses the source decimal type to check for overflow instead of the sink.
> 
> @bvarghese1 do you have a sense of how feasible it would be to check the output schema decimal type for overflow instead?
> 
> To me the CAST approach seems pretty reasonable because it's explicit. If we can adjust the overflow check I think it's a good option to consider.

@colinhicks Let me investigate this.

**Comment by spena** [comment]
@bvarghese1 Tip: I think you'll find easier to check during serialization if the conversion is allowed. The sum function will correctly do the sum, then the cast might change the schema (truncate it or something), then serializing to the Sink that has less precision/scale is the one that would fail.

**Comment by bvarghese1** [comment]
As @spena pointed out, the following CTAS doesn't preserve the sink type and uses the source type
```
CREATE TABLE S2 AS SELECT ID, CAST(sum(VALUE) AS DECIMAL(6,2)) AS SUM FROM DEC_TEST GROUP BY ID;
```
but, the following does:
```
CREATE TABLE S2 AS SELECT ID, sum(CAST(VALUE AS DECIMAL(6,2))) AS SUM FROM DEC_TEST GROUP BY ID;
```

**Comment by bvarghese1** [comment]
As dicussed with @spena and @colinhicks, we have decided to use the following syntax as workaround
```
CREATE TABLE S2 AS SELECT ID, sum(CAST(VALUE AS DECIMAL(6,2))) AS SUM FROM DEC_TEST GROUP BY ID;
```
The above syntax supports both upcasting and down casting and honors the users input in both cases. The sum udf will check for the up/down casted value in this scenario and raise a KSqlFunctionException.

**Pull request merged**